src/unity/toolkits/drawing_recognition/data_preparation.cpp

```diff
@@ -0,0 +1,376 @@
+    int FINAL_BITMAP_WIDTH = 28;
+    int FINAL_BITMAP_HEIGHT = 28;
+    float STROKE_WIDTH = 20.0;
+    int MAC_OS_STRIDE = 64;
+#ifdef __APPLE__
+    CGColorSpaceRef grayscale = CGColorSpaceCreateDeviceGray();
+    CGContextRef bitmap_context = CGBitmapContextCreate(
+        NULL, INTERMEDIATE_BITMAP_WIDTH, INTERMEDIATE_BITMAP_HEIGHT, 
+        8, 0, grayscale, kCGImageAlphaNone);
+    CGContextSetRGBStrokeColor(bitmap_context, 1.0, 1.0, 1.0, 1.0);
+    CGAffineTransform transform = CGAffineTransformIdentity;
+    CGMutablePathRef path = CGPathCreateMutable();
+    for (size_t i = 0; i < num_strokes; i++) {
+        const flex_list &stroke = simplified_drawing[i].get<flex_list>();
+        const flex_dict &start_point_dict = stroke[0].get<flex_dict>();
+        size_t num_points_in_stroke = stroke.size();
+        Point P; make_point(&P, start_point_dict);
+        // @TODO: When we make a dylib, we'll need some sophisticated 
+        // iOS/macOS checking here to figure out whether we need to 
+        // subtract 256 or not........
+        CGPathMoveToPoint(path, &transform, 
+            P.x, ((float)(INTERMEDIATE_BITMAP_WIDTH))-P.y);
+        for (size_t j = 1; j < num_points_in_stroke; j++) {
+            const flex_dict &point_dict = stroke[j].get<flex_dict>();
+            make_point(&P, point_dict);
+            CGPathAddLineToPoint(path, &transform, 
+                P.x, ((float)(INTERMEDIATE_BITMAP_WIDTH))-P.y);
+        }
+    }
+    CGContextSetLineWidth(bitmap_context, STROKE_WIDTH);
+    CGContextBeginPath(bitmap_context);
+    CGContextAddPath(bitmap_context, path);
+    CGContextStrokePath(bitmap_context);
+    CGPathRelease(path);
+    CGImageRef image = CGBitmapContextCreateImage(bitmap_context);
+    CGRect rectangle = CGRectMake(0.0, 0.0, 
+        ((float)(INTERMEDIATE_BITMAP_WIDTH)), 
+        ((float)(INTERMEDIATE_BITMAP_HEIGHT)));
+    CGImageRef cropped_image = CGImageCreateWithImageInRect(image, rectangle);
+    CGContextRef scale_28_context = CGBitmapContextCreate(NULL, 
+        FINAL_BITMAP_WIDTH, FINAL_BITMAP_HEIGHT, 8, 0, grayscale, 
+        kCGImageAlphaNone);
+    CGRect new_rect = CGRectMake(0.0, 0.0, 
+        ((float)(FINAL_BITMAP_WIDTH)), ((float)(FINAL_BITMAP_HEIGHT)));
+    CGContextDrawImage(scale_28_context, new_rect, cropped_image);
+    CGImageRef cropped_image_gray = CGBitmapContextCreateImage(
+        scale_28_context);
+    CFDataRef pixel_data = CGDataProviderCopyData(
+        CGImageGetDataProvider(cropped_image_gray));
+    const uint8_t* data_ptr = CFDataGetBytePtr(pixel_data);
+    uint8_t real_data[FINAL_BITMAP_WIDTH * FINAL_BITMAP_HEIGHT];
+    for (size_t row = 0; row < FINAL_BITMAP_WIDTH; row++) {
+        for (size_t col = 0; col < FINAL_BITMAP_HEIGHT; col++) {
+            // check platform here
+            real_data[row*FINAL_BITMAP_WIDTH+col] = (uint8_t)(
+                data_ptr[row * MAC_OS_STRIDE + col]);
+        }
+    }
+    return flex_image((const char *)real_data, 
```

Comment:
This function (and others that use CoreGraphics) looks like it might leak memory. All created CoreGraphics/CoreFoundation objects (like CGImageRef, CFDataRef, etc.) need to be released (CGImageRelease, CFRelease, etc.) to free their memory.

Take a look at the vega_renderer source code for examples of this. Let me know if you want help checking these codepaths in Instruments to see if they are leaking.

Reply:
That's a great point. Thank you for reminding me about that.